tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11458,48 +11499,38 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
     }
   }
 
-  // Validate that there is no recursion; start with the entry function.
-  // NOTE: the information gathered here could be used to bypass code generation
-  // on functions that are unreachable (as an early form of dead code
-  // elimination).
-  if (pEntryPointDecl) {
-    const auto *shaderModel =
-        hlsl::ShaderModel::GetByName(self->getLangOpts().HLSLProfile.c_str());
+  FunctionDecl *result = ValidateNoRecursion(self, pEntryPointDecl);
 
-    if (shaderModel->IsHS()) {
-      if (const HLSLPatchConstantFuncAttr *Attr =
-              pEntryPointDecl->getAttr<HLSLPatchConstantFuncAttr>()) {
-        NameLookup NL = GetSingleFunctionDeclByName(
-            self, Attr->getFunctionName(), /*checkPatch*/ true);
-        if (!NL.Found || !NL.Found->hasBody()) {
-          unsigned id =
-              Diags.getCustomDiagID(clang::DiagnosticsEngine::Level::Error,
-                                    "missing patch function definition");
-          Diags.Report(id);
-          return;
-        }
-        pPatchFnDecl = NL.Found;
+  if (result) {
+    self->Diag(result->getSourceRange().getBegin(), diag::err_hlsl_no_recursion)
+        << 0 << result->getName();
+  }
+
+  const auto *shaderModel =
+      hlsl::ShaderModel::GetByName(self->getLangOpts().HLSLProfile.c_str());
+
+  if (shaderModel->IsHS()) {
+    if (const HLSLPatchConstantFuncAttr *attr =
+            pEntryPointDecl->getAttr<HLSLPatchConstantFuncAttr>()) {
+      NameLookup NL = GetSingleFunctionDeclByName(self, attr->getFunctionName(),
+                                                  /*checkPatch*/ true);
+      if (!NL.Found || !NL.Found->hasBody()) {
+        self->Diag(attr->getLocation(),
+                   diag::err_hlsl_missing_patch_constant_function)
+            << attr->getFunctionName();
       }
+      pPatchFnDecl = NL.Found;
     }
+  }
 
-    hlsl::CallGraphWithRecurseGuard CG;
-    CG.BuildForEntry(pEntryPointDecl);
-    Decl *pResult = CG.CheckRecursion(pEntryPointDecl);
-    if (pResult) {
-      unsigned id =
-          Diags.getCustomDiagID(clang::DiagnosticsEngine::Level::Error,
-                                "recursive functions not allowed");
-      Diags.Report(pResult->getSourceRange().getBegin(), id);
-    }
-    if (pPatchFnDecl) {
-      CG.BuildForEntry(pPatchFnDecl);
-      Decl *pPatchFnDecl = CG.CheckRecursion(pEntryPointDecl);
-      if (pPatchFnDecl) {
-        unsigned id = Diags.getCustomDiagID(
-            clang::DiagnosticsEngine::Level::Error,
-            "recursive functions not allowed (via patch function)");
-        Diags.Report(pPatchFnDecl->getSourceRange().getBegin(), id);
-      }
+  if (pPatchFnDecl) {
+    FunctionDecl *patchResult = ValidateNoRecursion(self, pPatchFnDecl);
```

Comment:
There is an additional check we should have for patch constant / hull shader function case.  Neither function should be reachable from the other function.  In other words, the patch constant function must not be in the call graph under the hull shader entry function, and the hull shader entry function must not be in the call graph under the patch constant function.

Reply:
Perfect, I'll add a diagnostic for that and introduce CheckReachability.